bootstrap.py intermittent failure when fetching from Hackage

Increasingly, we've been seeing this lately:

```
Fetching http://hackage.haskell.org/package/os-string-2.0.7/revision/0.cabal...
Traceback (most recent call last):
  File "/usr/lib/python3.12/http/client.py", line 579, in _get_chunk_left
    chunk_left = self._read_next_chunk_size()
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/http/client.py", line 546, in _read_next_chunk_size
    return int(line, 16)
           ^^^^^^^^^^^^^
ValueError: invalid literal for int() with base 16: b'\r\n'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.12/http/client.py", line 595, in _read_chunked
    while (chunk_left := self._get_chunk_left()) is not None:
                         ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/http/client.py", line 581, in _get_chunk_left
    raise IncompleteRead(b'')
http.client.IncompleteRead: IncompleteRead(0 bytes read)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/runner/work/cabal/cabal/bootstrap/bootstrap.py", line 504, in <module>
    main()
  File "/home/runner/work/cabal/cabal/bootstrap/bootstrap.py", line 445, in main
    fetch_from_plan(plan, rootdir)
  File "/home/runner/work/cabal/cabal/bootstrap/bootstrap.py", line 355, in fetch_from_plan
    shutil.copyfileobj(resp, output_path.open('wb'))
  File "/usr/lib/python3.12/shutil.py", line 203, in copyfileobj
    while buf := fsrc_read(length):
                 ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/http/client.py", line 473, in read
    return self._read_chunked(amt)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/http/client.py", line 607, in _read_chunked
    raise IncompleteRead(b''.join(value)) from exc
http.client.IncompleteRead: IncompleteRead(3298 bytes read)
Error: Process completed with exit code 1.
```

(example: https://github.com/haskell/cabal/actions/runs/15705524267/job/44250333302?pr=10988)

--- a/bootstrap/bootstrap.py
+++ b/bootstrap/bootstrap.py
@@ -24,6 +24,8 @@ import subprocess
 import sys
 import tempfile
 import urllib.request
+import http.client
+import time
 from textwrap import dedent
 from typing import Optional, Dict, List, Tuple, \
                    NewType, BinaryIO, NamedTuple

--- a/bootstrap/bootstrap.py
+++ b/bootstrap/bootstrap.py
@@ -248,13 +248,6 @@ def install_sdist(dist_dir: Path, sdist_dir: Path, ghc: Compiler, flags: List[st
     # This should be fine for cabal-install dependencies.
     check_call([str(ghc.ghc_path), '--make', '-package-env=-', '-i', f'-odir={setup_dist_dir}', f'-hidir={setup_dist_dir}', '-o', setup, 'Setup'])
     check_call([setup, 'configure'] + configure_args)
-def download_with_requests(url, output_path):
-    with requests.get(url, stream=True, timeout=10) as r:
-        r.raise_for_status()
-        with output_path.open('wb') as f:
-            for chunk in r.iter_content(chunk_size=8192):
-                if chunk:
-                    f.write(chunk)
     check_call([setup, 'build'] + build_args)
     check_call([setup, 'install'] + build_dir)
 

--- a/bootstrap/bootstrap.py
+++ b/bootstrap/bootstrap.py
@@ -23,10 +23,10 @@ import shutil
 import subprocess
 import sys
 import tempfile
-import urllib.request
 from textwrap import dedent
 from typing import Optional, Dict, List, Tuple, \
                    NewType, BinaryIO, NamedTuple
+import requests
 
 #logging.basicConfig(level=logging.INFO)
 
@@ -136,6 +136,14 @@ class BadTarball(Exception):
             f'  found:    {self.found_sha256}',
         ])
 
+def download_with_requests(url, output_path):
+    with requests.get(url, stream=True, timeout=10) as r:
+        r.raise_for_status()
+        with output_path.open('wb') as f:
+            for chunk in r.iter_content(chunk_size=8192):
+                if chunk:
+                    f.write(chunk)
+
 def package_url(package: PackageName, version: Version) -> str:
     return f'http://hackage.haskell.org/package/{package}-{version}/{package}-{version}.tar.gz'
 
@@ -240,6 +248,13 @@ def install_sdist(dist_dir: Path, sdist_dir: Path, ghc: Compiler, flags: List[st
     # This should be fine for cabal-install dependencies.
     check_call([str(ghc.ghc_path), '--make', '-package-env=-', '-i', f'-odir={setup_dist_dir}', f'-hidir={setup_dist_dir}', '-o', setup, 'Setup'])
     check_call([setup, 'configure'] + configure_args)
+def download_with_requests(url, output_path):
+    with requests.get(url, stream=True, timeout=10) as r:
+        r.raise_for_status()
+        with output_path.open('wb') as f:
+            for chunk in r.iter_content(chunk_size=8192):
+                if chunk:
+                    f.write(chunk)
     check_call([setup, 'build'] + build_args)
     check_call([setup, 'install'] + build_dir)
 
@@ -351,8 +366,7 @@ def fetch_from_plan(plan : FetchPlan, output_dir : Path):
     sha = plan[path].sha256
     if not output_path.exists():
       print(f'Fetching {url}...')
-      with urllib.request.urlopen(url, timeout = 10) as resp:
-        shutil.copyfileobj(resp, output_path.open('wb'))
+      download_with_requests(url, output_path)
     verify_sha256(sha, output_path)
 
 def gen_fetch_plan(info : BootstrapInfo) -> FetchPlan :

--- a/bootstrap/bootstrap.py
+++ b/bootstrap/bootstrap.py
@@ -23,10 +23,10 @@ import shutil
 import subprocess
 import sys
 import tempfile
+import urllib.request
 from textwrap import dedent
 from typing import Optional, Dict, List, Tuple, \
                    NewType, BinaryIO, NamedTuple
-import requests
 
 #logging.basicConfig(level=logging.INFO)
 
@@ -136,13 +136,19 @@ class BadTarball(Exception):
             f'  found:    {self.found_sha256}',
         ])
 
-def download_with_requests(url, output_path):
-    with requests.get(url, stream=True, timeout=10) as r:
-        r.raise_for_status()
-        with output_path.open('wb') as f:
-            for chunk in r.iter_content(chunk_size=8192):
-                if chunk:
-                    f.write(chunk)
+def download_with_retry(url, output_path, retries=3, delay=2):
+    for attempt in range(retries):
+        try:
+            with urllib.request.urlopen(url, timeout=10) as resp:
+                with output_path.open('wb') as out_file:
+                    shutil.copyfileobj(resp, out_file)
+            return  # success
+        except http.client.IncompleteRead as e:
+            print(f"IncompleteRead error (attempt {attempt + 1}): {e}")
+        except Exception as e:
+            print(f"Other error (attempt {attempt + 1}): {e}")
+        time.sleep(delay)
+    raise Exception(f"Failed to download after {retries} attempts.")
 
 def package_url(package: PackageName, version: Version) -> str:
     return f'http://hackage.haskell.org/package/{package}-{version}/{package}-{version}.tar.gz'

--- a/bootstrap/bootstrap.py
+++ b/bootstrap/bootstrap.py
@@ -365,7 +365,7 @@ def fetch_from_plan(plan : FetchPlan, output_dir : Path):
     sha = plan[path].sha256
     if not output_path.exists():
       print(f'Fetching {url}...')
-      download_with_requests(url, output_path)
+      download_with_retry(url, output_path)
     verify_sha256(sha, output_path)
 
 def gen_fetch_plan(info : BootstrapInfo) -> FetchPlan :

